 <center> <h1>Fast Fourier Transform in Spartial Domain</h1> </center>
 <center> <h3>Saheed Adisa Ganiyu</h3> </center>
 
 <center> <h3>October, 2022.</h3> </center>

#### Motivation
Is there a faster way that will beat Karatsuba Algorithm for high degree polynomial multiplication?

<center><h2><span style="color:blue">Yes!</span></h2></center>

<center><h3><span style="color:blue">It is called Fast Fast Fourier Transform (FFT) algorithm(s)</span></h3></center>

### FOURIER TRANSFORM
#### DEFINITION
Let $f: \mathbb{R} \rightarrow \mathbb{C}$ be an integrable function, the Fourier transform of $f$ is a function $\hat{f}: \mathbb{R} \rightarrow \mathbb{C}$ defined by the formula
$$
\hat{f}(T):=\int_{-\infty}^{\infty} f(x) e^{-i 2 \pi x T} d x
$$

### INVERSE FOURIER TRANSFORM
#### DEFINITION
If $F: \mathbb{R} \rightarrow \mathbb{C}$ is an integrable function, then the inverse Fourier transform of $F$ is defined as
$$
\check{F}(x):=\int_{-\infty}^{\infty} F(T) e^{i 2 \pi x T} d T
$$

### DISCRETE FOURIER TRANSFORM
#### DEFINITION
Discrete Fourier transform of a finite sequence $\left(f_0, \ldots, f_{N-1}\right)$, is a sequence $\hat{f}=\left(\hat{f}_0, \ldots, \hat{f}_{N-1}\right)$, where
<math>
\begin{align}
\hat{f}_n:=\sum_{k=0}^{N-1} f_k e^{-i 2 \pi k n / N}
\end{align}
<math>

### DISCRETE INVERSE FOURIER TRANSFORM
#### DEFINITION
Discrete inverse Fourier transform of a sequence $F_0, \ldots, F_{N-1}$, is
<math>\begin{align}
\check{F}_k:=\frac{1}{N} \sum_{n=0}^{N-1} F_n e^{i 2 \pi k n / N} \text { for } k \in\{0, \ldots, N-1\} .
\end{align}<math>

### OBSERVATION (Primitive ROOT OF UNITY)
Take $\omega=e^{i \cdot \frac{2 \pi}{N}} \in \mathbb{C}$, then
+ $\omega^N=e^{2 \pi i}=1$,
* $\omega^k \neq 1$ for all $0<k<N$.

In [21]:
def root_of_unity( n, K = CC ) :
    """
    This function computes the primitive root of unity of degree `n` in `\\mathbb{C}`
    """
    return K( cos(2*pi/n) + i*sin(2*pi/n) )

def split_list( L ) :
    """
    Given a list `L` this function constructs lists `E` with elements of `L` having even indices \
    and `O` with elements of `L` having odd indices.
    """
    n = len(L)
    E = [ L[j] for j in [0,2,..,n-1] ]
    O = [ L[j] for j in [1,3,..,n-1] ]
    return E,O

def next_pow_2( n ) :
    """
    This function computes `2^N` s.t. `2^{N-1} < n \\leq 2^N`
    """
    assert n > 0, "n must be positive!"
    N = ZZ(ceil(log(n, 2)))
    return 2^N

<>:2: DeprecationWarning: invalid escape sequence \m
<>:18: DeprecationWarning: invalid escape sequence \l
<>:2: DeprecationWarning: invalid escape sequence \m
<>:18: DeprecationWarning: invalid escape sequence \l
<ipython-input-21-00672f265541>:2: DeprecationWarning: invalid escape sequence \m
  """
<ipython-input-21-00672f265541>:18: DeprecationWarning: invalid escape sequence \l
  """


In [22]:
#Verification
N=4
root_of_unity(N)^N

1.00000000000000

### GENERALIZED DISCRETE FOURIER TRANSFORM
+ Let $R$ be a ring and $\omega \in R$ be a primitive root of unity of degree $N$. The discrete Fourier transform **(DFT)** of a finite sequence $f=\left(f_0, \ldots, f_{N-1}\right)$ of length $N$ consisting of elements from $R$ is a sequence $\hat{f}=\left(\hat{f}_0, \ldots, \hat{f}_{N-1}\right)$ defined by the formula
$$
\hat{f}_n:=\sum_{k=0}^{N-1} f_k \omega^{-k n}
$$
\
+ The discrete inverse Fourier transform **(DIFT)** of a sequence $F=\left(F_0, \ldots, F_{N-1}\right)$ is a sequence $\check{F}=\left(\check{F}_0, \ldots, \check{F}_{N-1}\right)$ given by the formula
$$
\check{F}_k:=\frac{1}{N} \sum_{n=0}^{N-1} F_n \omega^{k n}
$$

In [23]:
#Computing fft using above definition
L = [1..4]; show("f="+latex(f))
N = len(L); #show("len(f)="+latex(N))
omega = root_of_unity(N,SR);
ff = [ sum( [ L[k]*omega^(-k*n) for k in range(N)] ) for n in range(N)]; 
show("\\hat{f}="+latex(ff))

f= 4x^{3} + 3x^{2} + 2x + 1

\hat{f}= \left[10, 2 i - 2, -2, -2 i - 2\right]

#### PROPOSITION
Let $R$ be a ring and $\omega \in R$ be a primitive root of unity of degree $N$. Then for any sequence $f$ of length $N$ the following identity holds
$$
f=\left(f^{\wedge}\right)^{\vee} .
$$

#### Thoerem (DFT of Polynomial)
- $R$ ring
- $f \in R[x]$ polynomial of degree $N-1$
- $\omega \in R$ primitive root of unity of degree $N(=1+\operatorname{deg} f)$
Then
$$
\hat{f}=\left(f(1), f(1 / \omega), \ldots, f\left(1 / \omega^{N-1}\right)\right) .
$$

In [24]:
#Computing the discrete Fourier transform (of the same polynomial as above) using the above theorem.
P.<x>=ZZ[];
f = P(L); show(f)
F = [f(omega^(j)) for j in [0..f.degree()] ]
show("\\hat{f}="+latex(F))

4*x^3 + 3*x^2 + 2*x + 1

\hat{f}= \left[10, -2 i - 2, -2, 2 i - 2\right]

<center><h2> Can we compute DFT and DIFT in a faster way that will beat Karatsuba Algorithm ? </h2><center>

<center><h2><span style="color:blue">Yes!</span></h2></center>

<center><h3><span style="color:blue">It is called Fast Fast Fourier Transform (FFT) algorithm(s)</span></h3></center>

### Cooley-Tukey FFT Algorithm
$$
\begin{aligned}
\hat{f}_j=\sum_{k=0}^{2^n-1} f_k \omega^{-k j}=\sum_{k=0}^{2^{n-1}-1} f_{2 k} \omega^{-2 k j} &+\sum_{k=0}^{2^{n-1}-1} f_{2 k+1} \omega^{-(2 k+1) j} \\
&=\sum_{k=0}^{2^{n-1}-1} f_{2 k}\left(\omega^2\right)^{-k j}+\omega^{-j} \cdot \sum_{k=0}^{2^{n-1}-1} f_{2 k+1}\left(\omega^2\right)^{-k j} .
\end{aligned}
$$

<b>The step-by-step</b> <a href="https://en.wikipedia.org/wiki/Cooley%E2%80%93Tukey_FFT_algorithm">Cooley-Tukey algorithm</a> to compute DFT of $(1,2,3,4)$.</p>
<math>
\begin{matrix} X_k= \underbrace{\sum \limits_{m=0}^{N/2-1} x_{2m}   e^{-\frac{2\pi i}{N/2} mk}}_{\mathrm{DFT\;of\;even-indexed\;part\;of\;} x_n} {} +  e^{-\frac{2\pi i}{N}k}
 \underbrace{\sum \limits_{m=0}^{N/2-1} x_{2m+1} e^{-\frac{2\pi i}{N/2} mk}}_{\mathrm{DFT\;of\;odd-indexed\;part\;of\;} x_n} =  E_k + e^{-\frac{2\pi i}{N}k} O_k.
\end{matrix}
</math>

#### Cooley-Tukey FFT algorithm
- $R$ is a ring
- $\omega \in R$ a primitive root of unity of degree $N$
- $N=2^n$ for some $n \in \mathbb{N}$\
INPUT: $f=\left(f_0, \ldots, f_{N-1}\right)$ a finite sequence of length $N$\
OUTPUT: the discrete Fourier transform $\hat{f}$ of $f$

1) If $n=0$, then return $\hat{f}=f$ and quit;\
2) take
$f_{\text {even }}:=\left(f_0, f_2, \ldots, f_{N-2}\right) \quad$ and $\quad f_{\text {odd }}:=\left(f_1, f_3, \ldots, f_{N-1}\right)$;\
3) compute recursively Fourier transforms
$\hat{f}_{\text {even }}=\left(\hat{e}_0, \ldots, \hat{e}_{N / 2-1}\right) \quad$ and $\quad \hat{f}_{\text {odd }}=\left(\hat{o}_0, \ldots, \hat{o}_{N / 2-1}\right) ;$\
4) combine the results taking for $j \in\{0, \ldots, N / 2-1\}$ :
$$
\hat{f}_j:=\hat{e}_j+\omega^{-j} \hat{o}_j \quad \text { and } \quad \hat{f}_{j+N / 2}:=\hat{e}_j-\omega^{-j} \hat{o}_j \text {; }
$$
5 return $\hat{f}=\left(\hat{f}_0, \ldots, \hat{f}_{n-1}\right)$

#### OBSERVATION
$~~~~~~~~~~~~~~~~~~~$ The time complexity of Cooley-Tukey algorithm is $\mathcal{O}(N \log N)$.

In [25]:
def fft( L, omega = 0 ) :
    """
    This function computes FFT of a list `L`, using Cooley-Tukey algorithm. \
    The length of the list `L` must be a power of 2.
    
    INPUT:
        
        - `L` list
        
        - `omega` - primitive root of unity of degree `|L|`.
        If `omega` is omited, it will be computed.
    """
    # get the length of L
    n = len(L)
    # stop the recursion if the list consists of a single element
    if n == 1 :
        return L
    assert NN(n).is_power_of(2), "The length of L must be a power of 2"
    # compute omega if it was not provided
    if omega == 0:
        omega = root_of_unity(n)

    # split L into its even and odd parts
    e, o = split_list(L)

    # compute FFTs of both parts 
    E = fft(e, omega^2)
    O = fft(o, omega^2)
    # recombine the partial results
    L0 = [ E[j] + omega^(-j)*O[j] for j in [0..n//2-1] ] 
    L1 = [ E[j] - omega^(-j)*O[j] for j in [0..n//2-1] ] 

    # return the final result
    return L0 + L1

In [26]:
# verification
res=fft( [1..4] )
show(res)

[10.0000000000000,
 -2.00000000000000 + 2.00000000000000*I,
 -2.00000000000000,
 -2.00000000000000 - 2.00000000000000*I]

<center> <b> <font size="+1">Flow Diagram</font> </b> </center> 
<img src="pic1.png" width="700" height="700"> <br>

#### DEFINITION
The convolution of two integrable functions $g$ and $f$ is a new function $f * g$ defined by the formula
$$
(f * g)(x):=\int_{-\infty}^{\infty} f(\xi) \cdot g(x-\xi) d \xi.
$$



![alt text](Conv_fft.png "Convolution Flow Graph")

#### CONVOLUTION
Definition\
■ $R$ ring\
■ $f=\left(f_0, \ldots, f_m\right)$ and $g=\left(g_0, \ldots, g_n\right)$ two finite sequences of elements of $R$ Then the sequence $(f \hat{*} g)_{0 \leq k \leq m+n}$ defined by the formula
$$
(f \hat{*} g)_k:=\sum_{j=\max \{0, k-n\}}^{\min \{k, m\}} f_j \cdot g_{k-j}
$$
is called the discrete convolution of $f$ and $g$.

#### POLYNOMIAL MULTIPLICATION VIA DFT
COROLLARY
If $f, g \in K[x]$ are two polynomials, then
$$
f * g=(\hat{f} \cdot \hat{g})^{\vee} .
$$
Here $*$ is the polynomial multiplication and $\cdot$ is the point-wise multiplication of their Fourier transforms.

Hence, the complexity of the polynomial multiplication using the above corollary is $\mathcal{O}(n+$ complexity of discrete Fourier transform $)$

#### FAST POLYNOMIAL MULTIPLICATION
- $R, f, g, N, \omega$\
OUTPUT: product $h:=f \cdot g$\
1) pad the lists to length $N$ :
$$
f:=\left(f_0, f_1, \ldots, f_m, 0, \ldots, 0\right), \quad g:=\left(g_0, g_1, \ldots, g_n, 0, \ldots, 0\right)
$$
2) compute DFTs:
$$
\hat{f}=\left(\hat{f}_0, \hat{f}_1, \ldots, \hat{f}_{N-1}\right), \quad \hat{g}=\left(\hat{g}_0, \hat{g}_1, \ldots, \hat{g}_{N-1}\right)
$$
3) multiply $\hat{f}$ and $\hat{g}$ coefficient-wisely:
$$
H:=\left(\hat{f}_0 \cdot \hat{g}_0, \ldots, \hat{f}_{N-1} \cdot \hat{g}_{N-1}\right)
$$
4) compute DIFT $h:=\check{H}$;\
5) return $h$.

#### FAST POLYNOMIAL MULTIPLICATION: EXAMPLE
- Take two polynomials:\
$$
f=4 x^3+3 x^2+2 x+1, \quad g=3 x^3+4 x^2+x+2,
$$
- Write their lists of coefficients padded with zeros:
$$
\begin{aligned}
    f &= [1, 2, 3, 4, 0, 0, 0, 0]\\
    g &= [2, 1, 4, 3, 0, 0, 0, 0]
\end{aligned}
$$
- Compute DFT's of both getting:
$$
\begin{aligned}
&\hat{\mathfrak{f}}=(10 .,-0.41-7.2 i,-2.0+2.0 i, 2.4-1.2 i,-2.0,2.4+1.2 i,-2.0-2.0 i,-0.41+7.2 i), \\
&\hat{\mathfrak{g}}=(10 ., 0.59-6.8 i,-2.0+2.0 i, 3.4+1.2 i, 2.0,3.4-1.2 i,-2.0-2.0 i, 0.59+6.8 i) .
\end{aligned}
$$
- Multiply coefficient-wisely:
$$
H=\left(100 .,-50 .-1.4 i,-1.8 \times 10^{-15}-8.0 i, 9.7-1.4 i,-4.0,9.7+1.4 i,-1.8 \times 10^{-15}+8.0 i,-50 .+1.4 i\right) \text {. }
$$
- Finally compute the discrete inverse Fourier transform getting
$$
h=12 x^6+25 x^5+22 x^4+22 x^3+12 x^2+5 x+2 .
$$

In [27]:
𝑓 = 4*x**3+3*x**2+2*x+1; show(f)
𝑔 = 3*x**3+4*x**2+x+2; show(g)

4*x^3 + 3*x^2 + 2*x + 1

3*x^3 + 4*x^2 + x + 2

In [28]:
def fast_poly_mult(f,g):
    """
    This function compute mutilplication of two polynomials.
    """
    P.<x> = ZZ[]
    # find the length of the resulting sequence
    N = 2*max( next_pow_2(f.degree()), next_pow_2(g.degree()))
    # convert f, g into lists of length N
    f_ = f.list() + [0]*(N - f.degree() - 1)
    g_ = g.list() + [0]*(N - g.degree() - 1)
    # compute FFT of both lists
    omega = root_of_unity(N)
    #K.<omega> = CyclotomicField(N)
    F = fft(f_, omega)
    G = fft(g_, omega)
    # Multiply the resulting sequences coordinate-wisely.
    H = [ F[j]*G[j] for j in [0..N-1] ]
    # the inverse transform = "forward" transform with omega replaced by 1/omega (remember about 1/N scaling)
    h_ = [ hi/N for hi in fft(H, omega^(-1)) ]
    # Convert the results back into integers
    h_ = [ round(real_part(hi)) for hi in h_ ]
    # cast onto a polynomial type
    h = P(h_)
    return h

In [29]:
res=fast_poly_mult(f,g)
show(res)

12*x^6 + 25*x^5 + 22*x^4 + 22*x^3 + 12*x^2 + 5*x + 2

In [30]:
# verification
res == f*g

True

### Comparison of execution time of each algorithms
Computing FFT's of sequences $(1,\dotsc, 2^k)$, for $k\leq 10$ using respectively: directly the definition, theorem about values of a polynomial, Cooley-Tukey algorithm. Compare the execution times of each method.</p>

In [31]:
# construct the sequences
seqs = [ [1..2^k] for k in [1..10] ]; #show(seqs)

In [19]:
%%timeit
# directly by definition

for l in seqs :
    omega = root_of_unity( len(l) )
    L = [ sum( l[k]*omega^(-k*n) for k in range(len(l)) ) for n in range(len(l)) ]

14 s ± 107 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [32]:
%%timeit
# using the theorem
P.<x> = ZZ[]

for l in seqs :
    f = P(l)
    omega = root_of_unity( f.degree() )
    F = [ f(omega^(-j)) for j in [0..f.degree()] ]

2.16 s ± 284 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [67]:
%%timeit
# using Cooley-Tukey

for l in seqs :
    F = fft(l)

227 ms ± 21.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


<!DOCTYPE html>
<html>
<style>
table, th, td {
  border:1px solid black;
}
</style>
<body>

<center> <b> <font size="+1">Summary</font> </b> </center>

<table style="width:100%">
  <tr>
    <th>Algorithms</th>
    <th>Execution Time</th>
  </tr>
  <tr>
    <td>Direct definition</td>
    <td>14.0 s</td>
  </tr>
  <tr>
    <td>By proposition</td>
    <td>2.16 s</td>
  </tr>
  <tr>
    <td>Cooley Tukey</td>
    <td>227 ms</td>
  </tr>
</table>

</body>
</html>

**Reference**
+ Przemysław Koprowski. Lectures on Computational Mathematics. 2021. http://www.pkoprowski.eu/lcm.\
+ Wikipedia, Cooley–Tukey FFT algorithm: https://en.wikipedia.org/wiki/Cooley%E2%80%93Tukey_FFT_algorithm
